storage/core/src/backend.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Low-level interface implemented by storage backends.
+
+pub use crate::{
+    info::{DbDesc, DbIndex},
+    Data,
+};
+
+/// Read-only database operations
+pub trait ReadOps {
+    /// Get value associated with given key.
+    fn get(&self, idx: DbIndex, key: &[u8]) -> crate::Result<Option<&[u8]>>;
+}
+
+/// Write database operation
+pub trait WriteOps {
+    /// Set value associated with given key.
+    fn put(&mut self, idx: DbIndex, key: Data, val: Data) -> crate::Result<()>;
+
+    /// Delete the value associated with given key.
+    fn del(&mut self, idx: DbIndex, key: &[u8]) -> crate::Result<()>;
+}
+
+/// Read-only transaction
+///
+/// If a clenup is required when the transaction closes, [Drop] should be implemented too.
+pub trait TxRo: ReadOps {}
```

Comment:
Why isn't there a `close()` or whatever method to end a transaction?

Same for TxRw transaction. There's no `revert()`.

Reply:
Since there is no guarantee that `close` or similar will be called by the user, I decided it's the job of the `Drop::drop` method to do the cleanup, as suggested in the comment. See the discussion on scoped transactions too, it's somewhat related. This applies to backend. The user-facing interface provides `abort` and `close` but these only consume the object and by dosing so `drop()` is triggered which would do the cleanup. This is also approach taken by some of the `lmdb` crates I was looking into.

Reply:
There's no guarantee, but the choice should probably be there. Now I can't say this must go in here, since this is the backend of the storage. As long as these methods exist for the storage frontend, I'm OK with that.

The way I do it, usually, is that I make drop call the close/revert functions. That way both possibilities exist.

Reply:
> The way I do it, usually, is that I make drop call the close/revert functions. That way both possibilities exist.

Then close/revert would have to take `&mut self` rather than `self` since drop has the signature of `fn drop(&mut self)`. So I think it's cleaner to just use `drop` for cleanup, which has to happen anyway in case the user does not call `abort()` or similar explicitly and avoid duplication of this functionality in the backend code.

The high-level interface includes the `TransactionRo::close` and `TransactionRw::abort` methods but these are merely convenient ways to call `std::mem::drop` early with a more accurate name.

Reply:
That's fair. I forgot about moving self.